### ISSUE TITLE
device not added to ACL if already approved

### ISSUE THREAD

**Issue by dpocock** [intel/thunderbolt-software-user-space#34]
My Thunderbolt 3 dock was already connected when my system booted up.  Running the command `tbtadm devices` showed that it was `authorized` but not in ACL.

If I try to add it to the ACL by manually typing `tbtadm approve 0-1` it would complain that it was already approved.

I had to detach the dock and re-attach it and run the `approve` command again to really get it into the ACL.

I'm using version 0.9.2-1 on Debian

**Comment by ybernat** [comment]
This was intentional before we had `--once`, so it's possible to approve a device, remove it from ACL, and it'll not be added again just because the user wanted to approve another device.

We can reconsider it now or maybe add a separated command for adding a device to ACL.

Suggestions?

**Comment by dpocock** [comment]
maybe add a `--force` option, then it may also be a good idea to expand the error message from the `approve` command so it says something like `Device already approved but not in ACL, please use --force option`

**Comment by ybernat** [comment]
I'm just not sure if the meaning is clear enough. Wouldn't the user expect something like `add` for adding to ACL instead of using `approve` on already approved device?

**Comment by dpocock** [comment]
That could be better, but then you have situations where `approve` is also modifying the ACL so maybe `approve-once` should become the default `approve`?

**Comment by superm1** [comment]
@dpocock can you please confirm your BIOS settings related to Thunderbolt?  Do you have pre-boot modules or boot support checked?  (At least on Dell machines the default is that pre-boot modules and boot support both are NOT checked).
If the device is working in the OS without authorizing and you don't have that checked then we're looking at a bug on the Linux side in the thunderbolt kernel module, possibly related to your machine.

I just double checked with both Windows 10 and 4.15-rc1 and a TB3 dock plugged in "before" boot and those unchecked.  The dock isn't authorized automatically in Windows or Linux.

If I check those check boxes and reboot then I get a popup in Windows:
"Device was approved during boot but is not part of the current approved devices list"
but it does work.

**Comment by dpocock** [comment]
Yes, this is a Thinkpad and Lenovo Thunderbolt 3 dock and the BIOS setting for pre-boot had been enabled.

It would be really useful if this package included a brief document about these options and their security implications as many people will be encountering this for the first time when they buy Thunderbolt hardware.  There is already some [really bad advice available as a 'solution' on the Lenovo forum](https://forums.lenovo.com/t5/Linux-Discussion/ThunderBolt-3-Dockingstation-and-Linux/m-p/3745070#M9694)

**Comment by superm1** [comment]
The best place to tell someone about the implications of turning on a BIOS option is in BIOS setup :)

I think something similar to what Windows is doing to notify the user that pre-boot support is enabled would be a good idea.  I think it can be done with a poor man's GUI too.

@ybernat how about something like this:
1. udev rule that matches on devices that were pre-authorized.
2. udev rule launches something for tbtadm to examine the situation and determine if they were in the ACL already. 
3. If not in the ACL already use `xdg-notify` with a similar message to the Windows message?

**Comment by finikorg** [comment]
I think udev rule matching pre-authorized device and adding (with the help of some tool) them to acl would be best solution IMO.

**Comment by ybernat** [comment]
@dpocock
If you worry about a race-condition, I don't think this is something to worry about.
1. The user shouldn't run the 2 commands in parallel.
2. Even if the user does (or in multi-user environment it happen somehow), the `add` command is typically done for already authorized device and the `approve`/`approve-all` command handles only unauthorized devices, as you have noticed.

@superm1
Good point.
1. Do you know by any chance if in this case the SW gets 'approved on boot' flag in the 'device connected' notification from CM? (Can be tested with Windows if you open the SW logs, it should print there this information when the service gets the notification.)
2. This will need a kernel patch to expose this information in sysfs. @westeri, your opinion?

@finikorg
So your suggestion is not to warn the user but to auto-add it to ACL?
I'm not sure this is the best option. It makes sense if it's in the FW ACL, but it makes less sense if this is from BIOS setting that allows everything to connect (and we can't tell the difference easily, unless @westeri agrees to add another change in the driver to expose the FW ACL). At least this is how I see it.

**Comment by ybernat** [comment]
Oh, and about the documentation, the kernel documentation already mentions why the user-approval is important: https://www.kernel.org/doc/html/v4.13/admin-guide/thunderbolt.html#security-levels-and-how-to-use-them

@dpocock 
Where did you expected such a documentation to appear, in the man page for `tbtadm`?
Just bear in mind that all that we can add is a generic description. We can't document all the different BIOS settings from all the different vendors. Each one has a different set of settings and different behavior regarding such things like "allow on boot" etc.

**Comment by westeri** [comment]
If it really brings some value to the user, then we may expose it (approved on boot) one way or another. I'm trying to keep sysfs attributes at minimum but from time to time we may need to add more.

**Comment by finikorg** [comment]
Maybe we should rename tbtadm command line tool:
approve = approve --once => authorize device
enroll = approve => authorize and add to acl (if device is authorized just add to acl)

**Comment by ybernat** [comment]
@finikorg 
I see it as an API change. We need a good reason to do it.
Anyway, as I mentioned in your pull request, in SL2 we can't add to ACL when the device is already approved, which is a good reason to go with a separated command like `add` and return an error if it's SL2 (because then, `approve` must be used; there is no add to ACL w/o approving the device in this case).

**Comment by superm1** [comment]
@ybernat It isn't immediately obvious to me where the SW logs are for Windows.  Could you notify me where they are?
I checked in event viewer and `C:\program files (x86)\intel\thunderbolt software`.  Didn't see anything that looked like a log.

**Comment by ybernat** [comment]
@superm1 
Sorry, I keep forgetting this isn't a common knowledge...
It uses ETW mechanism. The easiest way is to use [MessageAnalyzer](https://www.microsoft.com/en-us/download/details.aspx?id=44226), open a 'Live trace' session and choose intel-thunderbolt-* components in the provider list (the most important in this case is the service).

**Comment by superm1** [comment]
@ybernat 
Thanks.  I tried to set up a live trace with Intel-Thunderbolt-Service as a provider and hit start.
I'm getting this:

> Enable session "MM-ETW-Livecapture-$GUID" failed

Any ideas?  Do I need to do something else too when configuring?

**Comment by superm1** [comment]
@ybernat I got it running as admin, but it seems like I'm not going to have this running early enough at boot to catch what you're interested in since it's a live capture.  Should I just restart thunderbolt service and watch for data?

**Comment by superm1** [comment]
@ybernat OK so with restarting the service here's what I caught.
1. There was a device connected notification for both the dock and the cable.  
There are a ton of debugging events related to that, not sure what you'd be interested in.

**Comment by ybernat** [comment]
@superm1 
Yes, restarting the service is the simplest solution for this case. It'll get all the "device connected" notifications again so we can see if the device is marked as approved on boot.
(Another solution is to use [AutoLogger](https://msdn.microsoft.com/en-us/library/aa363687.aspx) so the logging starts on boot, but this is a more complicated solution and not needed here).

**Comment by ybernat** [comment]
@superm1 
After a DEVICE_CONNECTED_NOTIFICATION there is a line similar to:

> Device details: Certified - 1, Device power - 0, Device security level - 2, First depth device - 0, Approved bit - 0, Boot - 0

We expect here 'Approved bit' to be 1, of course, and the question is about the value of 'Boot'.

**Comment by superm1** [comment]
Yep, here you go:
```
188	None	2017-12-11T17:04:11.9183429				Intel_Thunderbolt_Service	Device connected notification (message 1/1): Controller PCI\VEN_8086&DEV_15D9&SUBSYS_11112222&REV_02\6&18f0d7b9&0&000000D8_0   -----> [File: ConnectionManager.cpp, Function: ConnectionManager::OnDeviceConnect, Line: 332].	
189	None	2017-12-11T17:04:11.9183476				Intel_Thunderbolt_Service	Device details: Certified - 1, Device power - 0, Device security level - 2, First depth device - 0, Approved bit - 1, Boot - 1   -----> [File: ConnectionManager.cpp, Function: ConnectionManager::OnDeviceConnect, Line: 343].	
```

```
216	None	2017-12-11T17:04:12.5532807				Intel_Thunderbolt_Service	FW Notification: DEVICE_CONNECTED_NOTIFICATION message 1 out of 1, Controller: PCI\VEN_8086&DEV_15D9&SUBSYS_11112222&REV_02\6&18f0d7b9&0&000000D8_0   -----> [File: MessageFromFwEvent.cpp, Function: MessageFromFwEvent::HandleFwToSwNotification, Line: 247].	
217	None	2017-12-11T17:04:12.5532961				Intel_Thunderbolt_Service	Device connected notification (message 1/1): Controller PCI\VEN_8086&DEV_15D9&SUBSYS_11112222&REV_02\6&18f0d7b9&0&000000D8_0   -----> [File: ConnectionManager.cpp, Function: ConnectionManager::OnDeviceConnect, Line: 332].	
218	None	2017-12-11T17:04:12.5533105				Intel_Thunderbolt_Service	Device details: Certified - 1, Device power - 0, Device security level - 2, First depth device - 0, Approved bit - 1, Boot - 1   -----> [File: ConnectionManager.cpp, Function: ConnectionManager::OnDeviceConnect, Line: 343].	
```

**Comment by ybernat** [comment]
Great. So it does show that the device was approved on boot, and thus we can warn the user and ask if it should be added to ACL.

**Comment by finikorg** [comment]
@ybernat So should we have "acl-add" or "add-acl" command for SL1?
For SL2 there is a check in the kernel in key_store() and if device is authorized it returns -EBUSY.

**Comment by ybernat** [comment]
As we already have `remove`, `add` makes more sense to me.

BTW, maybe we want to change the driver interface to allow adding to ACL in SL2 even after the device is already authorized. Comments?

**Comment by hamaadshah** [comment]
Hi All: I can see my eGPU as authorized however I cannot add it to the ACL. Any suggestions?

**Comment by westeri** [comment]
Did you try 'tbtadm add' command?

**Comment by hamaadshah** [comment]
Yes indeed. I tried the following

samson@samson-MacBookPro:~/Projects/thunderbolt-software-user-space/build$ sudo tbtadm add 0-1
Adding to ACL is not relevant in SL0

Any help on this would be appreciated - I'm not an expert in this particular area

**Comment by hamaadshah** [comment]
Yes indeed. I tried the following

samson@samson-MacBookPro:~/Projects/thunderbolt-software-user-space/build$ sudo tbtadm add 0-1
Adding to ACL is not relevant in SL0

Any help on this would be appreciated - I'm not an expert in this particular area

**Comment by westeri** [comment]
Well, you are on Apple system where the security level is set to none (it always authorizes devices). You don't need any of the ACL stuff. You may want to add 'intel_iommu=on' to your kernel command line to make sure VT-d (IOMMU) is enabled if you care about security on that system.

**Comment by hamaadshah** [comment]
Gotcha - thanks for that. On Mac OSX I've got the eGPU running neatly however on my Linux OS (on the same MacBook Pro) struggling to get it running. Most likely that running Linux OS on a MacBook Pro and expecting an eGPU to work nicely is a bit of an ask. 

I appreciate your time. Thank you.

**Comment by westeri** [comment]
Could you attach (or send directly to me) dmesg from the system when you plug in eGPU? I would like to check whether Linux PCI resource allocation works as expected. Other than that, I can't help much as I have very little knowledge how eGPUs are used in Linux.
